dali/core/mm/default_resources.cc

```diff
@@ -0,0 +1,248 @@
+#endif
+}
+
+inline std::shared_ptr<pinned_async_resource> CreateDefaultPinnedResource() {
+  using resource_type = mm::async_pool_resource<mm::memory_kind::pinned>;
+  auto upstream = std::make_shared<rmm::mr::pinned_memory_resource>();
+  auto rsrc = std::make_shared<resource_type>(upstream.get());
+  return make_shared_composite_resource(std::move(rsrc), std::move(upstream));
+}
+
+}  // namespace
+
+template <> DLL_PUBLIC
+const std::shared_ptr<host_memory_resource> &GetDefaultResource<memory_kind::host>() {
+  if (!g_resources.host)
+    g_resources.host = CreateDefaultHostResource();
+  return g_resources.host;
+}
+
+template <> DLL_PUBLIC
+const std::shared_ptr<pinned_async_resource> &GetDefaultResource<memory_kind::pinned>() {
+  if (!g_resources.pinned_async) {
+    static CUDARTLoader init_cuda;  // force initialization of CUDA before creating the resource
+    static auto cleanup = AtExit([] {
+      g_resources.ReleasePinned();
+    });
+    g_resources.pinned_async = CreateDefaultPinnedResource();
+  }
+  return g_resources.pinned_async;
+}
+
+template <> DLL_PUBLIC
+void SetDefaultResource<memory_kind::host>(shared_ptr<host_memory_resource> resource) {
+  g_resources.host = std::move(resource);
+}
+
+
+template <> DLL_PUBLIC
+void SetDefaultResource<memory_kind::host>(host_memory_resource *resource, bool own) {
+  SetDefaultResource<memory_kind::host>(wrap(resource, own));
+}
+
+
+template <> DLL_PUBLIC
+void SetDefaultResource<memory_kind::pinned>(shared_ptr<pinned_async_resource> resource) {
+  g_resources.pinned_async = std::move(resource);
+}
+
+
+template <> DLL_PUBLIC
+void SetDefaultResource<memory_kind::pinned>(pinned_async_resource *resource, bool own) {
+  SetDefaultResource<memory_kind::pinned>(wrap(resource, own));
+}
+
+
+const std::shared_ptr<device_async_resource> &GetDefaultDeviceResource(int device_id) {
+  if (device_id < 0) {
+    CUDA_CALL(cudaGetDevice(&device_id));
+  }
+  if (g_resources.device.empty()) {
+    int ndevs = 0;
+    CUDA_CALL(cudaGetDeviceCount(&ndevs));
+    g_resources.device.resize(ndevs);
+  }
+  if (static_cast<size_t>(device_id) >= g_resources.device.size())
+    throw std::out_of_range(make_string(device_id, " is not a valid CUDA device index."));
+  if (!g_resources.device[device_id]) {
+    DeviceGuard devg(device_id);
+    static CUDARTLoader init_cuda;  // force initialization of CUDA before creating the resource
+    static auto cleanup = AtExit([] {
+      g_resources.ReleaseDevice();
+    });
+    g_resources.device[device_id] = CreateDefaultDeviceResource();
+  }
+  return g_resources.device[device_id];
+}
+
+template <> DLL_PUBLIC
+const std::shared_ptr<device_async_resource> &GetDefaultResource<memory_kind::device>() {
+  return GetDefaultDeviceResource(-1);
+}
+
+void SetDefaultDeviceResource(int device_id, std::shared_ptr<device_async_resource> resource) {
+  int ndevs = 0;
+  CUDA_CALL(cudaGetDeviceCount(&ndevs));
```

Comment:
Shouldn't you check `g_resources.device.empty`.

Reply:
:+1:  I can.